sdk/core/azure-core/azure/core/messaging.py

```diff
@@ -0,0 +1,154 @@
+    :type id: Optional[str]
+    :keyword extensions: Optional. A CloudEvent MAY include any number of additional context attributes
+     with distinct names represented as key - value pairs. Each extension must be alphanumeric, lower cased
+     and must not exceed the length of 20 characters.
+    :type extensions: Optional[Dict]
+    :ivar source: Identifies the context in which an event happened. The combination of id and source must
+     be unique for each distinct event. If publishing to a domain topic, source must be the domain name.
+    :vartype source: str
+    :ivar data: Event data specific to the event type.
+    :vartype data: object
+    :ivar type: Type of event related to the originating occurrence.
+    :vartype type: str
+    :ivar time: The time (in UTC) the event was generated.
+    :vartype time: ~datetime.datetime
+    :ivar dataschema: Identifies the schema that data adheres to.
+    :vartype dataschema: str
+    :ivar datacontenttype: Content type of data value.
+    :vartype datacontenttype: str
+    :ivar subject: This describes the subject of the event in the context of the event producer
+     (identified by source).
+    :vartype subject: str
+    :ivar specversion: Optional. The version of the CloudEvent spec. Defaults to "1.0"
+    :vartype specversion: str
+    :ivar id: An identifier for the event. The combination of id and source must be
+     unique for each distinct event. If not provided, a random UUID will be generated and used.
+    :vartype id: str
+    :ivar extensions: A CloudEvent MAY include any number of additional context attributes
+     with distinct names represented as key - value pairs. Each extension must be alphanumeric, lower cased
+     and must not exceed the length of 20 characters.
+    :vartype extensions: Dict
+    """
+
+    def __init__(self, source, type, **kwargs):  # pylint: disable=redefined-builtin
+        # type: (str, str, **Any) -> None
+        self.source = source  # type: str
+        self.type = type  # type: str
+        self.specversion = kwargs.pop("specversion", "1.0")  # type: str
+        self.id = kwargs.pop("id", str(uuid.uuid4()))  # type: str
+        self.time = kwargs.pop("time", datetime.now(TZ_UTC))  # type: datetime
+        self.datacontenttype = kwargs.pop("datacontenttype", None)  # type: str
+        self.dataschema = kwargs.pop("dataschema", None)  # type: str
+        self.subject = kwargs.pop("subject", None)  # type: str
+        self.extensions = {}  # type: Dict
+        _extensions = dict(kwargs.pop("extensions", {}))
+        for key in _extensions.keys():
+            if not key.islower() or not key.isalnum():
+                raise ValueError(
+                    "Extension attributes should be lower cased and alphanumeric."
+                )
+        self.extensions.update(_extensions)
+        self.data = kwargs.pop("data", None)  # type: object
```

Comment:
It has been pointed out to me that `data` is optional, and that there is a difference between `data=None` and no `data`. Is this a correct statement? If so, we are technically *not* doing the right thing here.

Reply:
Yes, the CloudEvent spec actually makes a different with no data and `data=None`

Reply:
IOW, those two JSONs represent different CloudEvent:
```json
{
   "source": "whatever",
   "type": "whatever2",
}
```
```json
{
   "source": "whatever",
   "type": "whatever2",
   "data": null
}
```

Reply:
Given CloudEvent accepts extensions, it's possible user decides to pass data as extensions only